File: firebase/functions/lib/events/calendar/abstract_calendar_feed.dart
import 'dart:async';
import 'dart:io';

import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    hide CloudFunction;
import '../../cloud_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'package:data_models/community/community.dart';

/// Generic logic for returning a calendar feed. Functions that implement this should expect the
/// space id to be present as the second request path parameter, as in '/space/[space_id]/xyz'.
abstract class AbstractCalendarFeed implements CloudFunction {
  Future<String> generateData({required Community community});
  ContentType getContentType();

  Future<void> expressAction(ExpressHttpRequest expressRequest) async {
    try {
      if (expressRequest.requestedUri.pathSegments.length != 3) {
        throw Exception("Bad path");
      }
      final communityId = expressRequest.requestedUri.pathSegments[1];

      final communityDocs = await firestore
          .collection('community')
          .where(Community.kFieldDisplayIds, arrayContains: communityId)
          .get();
      if (communityDocs.isEmpty) {
        expressRequest.response.statusCode = HttpStatus.notFound;
        return expressRequest.response.close();
      }
      final community =
          Community.fromJson(communityDocs.documents.first.data.toMap());

      final data = await generateData(community: community);

      expressRequest.response.headers.contentType = getContentType();
      expressRequest.response.write(data);
    } catch (e, stacktrace) {
      print('Error during action');
      print(e);
      print(stacktrace);
      expressRequest.response.statusCode = HttpStatus.internalServerError;
    }
    await expressRequest.response.close();
  }

  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions.https.onRequest(expressAction);
  }
}

File: firebase/functions/src/events/calendar/abstract_calendar_feed.ts
import * as functions from 'firebase-functions';
import { CloudFunction } from '../../cloud_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';

export interface Community {
  id?: string;
  displayIds?: string[];
  name?: string;
  [key: string]: unknown;
}

export abstract class AbstractCalendarFeed implements CloudFunction {
  abstract readonly functionName: string;

  abstract generateData(params: { community: Community }): Promise<string>;
  abstract getContentType(): string;

  async expressAction(req: functions.https.Request, res: import('express').Response): Promise<void> {
    try {
      const segments = req.path.split('/').filter(Boolean);
      if (segments.length !== 2) {
        res.status(400).send('Bad path');
        return;
      }
      const communityId = segments[1];

      const communityDocs = await firestore
        .collection('community')
        .where('displayIds', 'array-contains', communityId)
        .get();

      if (communityDocs.empty) {
        res.status(404).send('Not found');
        return;
      }

      const community = firestoreUtils.fromFirestoreJson(communityDocs.docs[0].data()) as unknown as Community;
      const data = await this.generateData({ community });
      res.setHeader('Content-Type', this.getContentType());
      res.send(data);
    } catch (e) {
      console.error('Error during action', e);
      res.status(500).send(String(e));
    }
  }

  register(_functions: typeof functions): functions.HttpsFunction {
    return _functions.https.onRequest((req, res) => this.expressAction(req, res));
  }
}
